### PR TITLE
[Backport 6.0] mv: handle different ERMs for base and view table

### PR DESCRIPTION
**Pull request by wmitros**
When calculating the base-view mapping while the topology is changing, we may encounter a situation where the base table noticed the change in its effective replication map while the view table hasn't, or vice-versa. This can happen because the ERM update may be performed during the preemption between taking the base ERM and view ERM, or, due to f2ff701, the update may have just been performed partially when we are taking the ERMs.

Until now, we assumed that the ERMs are synchronized while calling finding the base-view endpoint mapping, so in particular, we were using the topology from the base's ERM to check the datacenters of all endpoints. Now that the ERMs are more likely to not be the same, we may try to get the datacenter of a view endpoint that doesn't exist in the base's topology, causing us to crash.

This is fixed in this patch by using the view table's topology for endpoints coming from the view ERM. The mapping resulting from the call might now be a temporary mapping between endpoints in different topologies, but it still maps base and view replicas 1-to-1.

Fixes: #17786
Fixes: #18709

(cherry-picked from 519317dc5833127732061c3eea2e2e99577b31e6)

This commit also includes the follow-up patch that removes the flakiness from the test that is introduced by the commit above. The flakiness was caused by enabling the
delay_before_get_view_natural_endpoint injection on a node and not disabling it before the node is shut down. The patch removes the enabling of the injection on the node in the first place.
By squashing the commits, we won't introduce a place in the commit history where a potential bisect could mistakenly fail.

Fixes: https://github.com/scylladb/scylladb/issues/18941

(cherry-picked from 0de3a5f3ff4b060227a7150aa90546920bc54ce6)

### CODE REVIEW

**Comment by wmitros** [comment]
The patch used in this backport (https://github.com/scylladb/scylladb/commit/0de3a5f3ff4b060227a7150aa90546920bc54ce6) is still only in `next` - we should wait until it promotes before merging this one.
I'm sending the PR already so that the CI finishes more quickly

**Review by piodul — APPROVED** [review]
LGTM, pre-approving (need to wait both for promotion and CI for this PR first).

**Comment by piodul** [comment]
There is clearly something wrong with CI job for PRs targeted towards 6.0.

I don't want to delay this further. I'm going to merge right now because:

- The commits this was backported from (519317d and 0de3a5f) passed promotion to master
- Master and 6.0 should still be quite similar at this point (at least with regards to materialized views code)
- The tests will still have to run for this in order to be promoted to branch-6.0

**Comment by yaronkaikov** [comment]
> There is clearly something wrong with CI job for PRs targeted towards 6.0.
> 
> I don't want to delay this further. I'm going to merge right now because:
> 
> * The commits this was backported from ([519317d](https://github.com/scylladb/scylladb/commit/519317dc5833127732061c3eea2e2e99577b31e6) and [0de3a5f](https://github.com/scylladb/scylladb/commit/0de3a5f3ff4b060227a7150aa90546920bc54ce6)) passed promotion to master
> * Master and 6.0 should still be quite similar at this point (at least with regards to materialized views code)
> * The tests will still have to run for this in order to be promoted to branch-6.0

@piodul Fonud the problem, it seems that one of my debug commits got into 6.0, this is why the CI is not working
Fixing it now

**Comment by yaronkaikov** [comment]
Running https://jenkins.scylladb.com/job/scylla-6.0/job/scylla-ci/30/

**Comment by piodul** [comment]
@yaronkaikov thanks. I already merged it by the time you restarted, but it's good to run it anyway (and thankfully it passed).

**Comment by piodul** [comment]
Landed on `branch-6.0` (https://github.com/scylladb/scylladb/commit/3c47ab985198398a38e5c8647d63b9e7f74eb488), closing.

**Pull request closed without merging**
